packages/components/src/Chip/Chip.tsx

```diff
@@ -31,29 +33,73 @@ export const Chip = ({
   const prefix = useChildComponent(children, d => d.type === Chip.Prefix);
   const suffix = useChildComponent(children, d => d.type === Chip.Suffix);
 
+  const [labelRef, labelFullyVisible] = useInView<HTMLSpanElement>();
+  const [headingRef, headingFullyVisible] = useInView<HTMLSpanElement>();
+  const tooltipMessage = getTooltipMessage(
+    labelFullyVisible,
+    headingFullyVisible,
+    label,
+    heading,
+  );
+
   return (
-    <button
-      className={classes}
-      onClick={(ev: React.MouseEvent<HTMLButtonElement>) =>
-        onClick && onClick(value, ev)
-      }
-      tabIndex={tabIndex}
-      onKeyDown={onKeyDown}
-      aria-label={ariaLabel || label}
-      disabled={disabled}
-      role={role}
-      type="button"
-    >
-      {prefix}
-      <Typography size="base" fontWeight="medium">
-        {heading}
-      </Typography>
-      {heading && <span className={styles.chipBar} />}
-      <Typography size="base">{label}</Typography>
-      {suffix}
-    </button>
+    <Tooltip message={tooltipMessage}>
+      <button
+        className={classes}
+        onClick={(ev: React.MouseEvent<HTMLButtonElement>) =>
+          onClick && onClick(value, ev)
+        }
+        tabIndex={tabIndex}
+        onKeyDown={onKeyDown}
+        aria-label={ariaLabel || label}
+        disabled={disabled}
+        role={role}
+        type="button"
+      >
+        {prefix}
+        <div className={styles.chipContent}>
+          {heading && (
+            <>
+              <Typography size="base" fontWeight="medium">
+                {heading}
+                <span ref={headingRef} />
+              </Typography>
+              <span className={styles.chipBar} />
+            </>
+          )}
+          <Typography size="base">
+            {label}
+            <span ref={labelRef} />
+          </Typography>
+          {!labelFullyVisible && (
+            <div
+              className={styles.truncateGradient}
+              data-testid="ATL-Chip-Truncation-Gradient"
+            />
+          )}
+        </div>
+        {suffix}
+      </button>
+    </Tooltip>
   );
 };
 
+function getTooltipMessage(
```

Comment:
Can we either turn this into a useMemo or flip the format to be a single return value? This is probably just my own personal style creeping in here though, so if the multiple return statements are a common pattern here feel free to ignore me.

Reply:
>turn this into a useMemo

do you mean to memo-ize the value of `tooltipMessage` in the component body where we call this? since this function itself is defined outside the component it's only being made the once

>multiple return statements are a common pattern here feel free to ignore me.

it's a fair question. I'm not sure if that's an established pattern or we have opinions on it in Atlantis. I personally don't mind them because I think they're quite readable, but that's also an opinion

what are you thinking for a single return value? something like

```
function getTooltipMessage() {
 let message = "";
 
  if (heading && !headingFullyVisible) {
     message = `${heading} | ${label}`;
  }

  if (!labelFullyVisible) {
    message = label;
  }

  return message;
}
```